File: Stack/Basic Calculator/BasicCalculator.rb
def calculate(s)
  output, curr, sign, stack = 0, 0, 1, []
  (0...s.length).each do |i|
    c = s[i]
    if c.match?(/\d/)
      curr = (curr * 10) + c.to_i
    elsif c.match?(/[+\-]/)
      output += curr * sign
      curr = 0
      sign = 1 if c == "+"
      sign = -1 if c == "-"
    elsif c == "("
      stack.push(output)
      stack.push(sign)
      output = 0
      sign = 1
    elsif c == ")"
      output += curr * sign
      output *= stack.pop
      output += stack.pop
      curr = 0
    end
  end
  output + (curr * sign)
end

File: Stack/Basic Calculator/BasicCalculator.go
func calculate(str string) int {
	output, curr, sign := 0, 0, 1
	stack := arraystack.New()
	for i := 0; i < len(str); i++ {
		c := str[i]
		if c >= '0' && c <= '9' {
			curr = (curr * 10) + int(c-'0')
		} else if c == '+' || c == '-' {
			output += curr * sign
			curr = 0
			if c == '+' {
				sign = 1
			} else {
				sign = -1
			}
		} else if c == '(' {
			stack.Push(output)
			stack.Push(sign)
			output = 0
			sign = 1
		} else if c == ')' {
			output += curr * sign
			prevSign, _ := stack.Pop()
			prevOutput, _ := stack.Pop()
			output *= prevSign.(int)
			output += prevOutput.(int)
			curr = 0
		}
	}
	return output + (curr * sign)
}